### ISSUE TITLE
Exception occurred processing WSGI script - Internal Server Error - Daily Crash

### ISSUE THREAD

**Issue by birwin79** [open-oni/open-oni#607]
Hello,

My ONI docker environment based on ONI v1.1.0 regularly crashes once a day.  

It starts when the visitor gets the generic "internal server error" in the browser.  After rapidly refreshing/reloading the page 3 or 4 times, the application returns independently and resumes regular operation.  It is stable for several hours and does not crash during use, only if it is idle for some time.

I'm trying to narrow it down, but I get a different error code in the log file at each page refresh until it comes back up.  
Restoration of service usually occurs after the third or fourth reload in the browser.
I'm running on a server with AMD 3950x 16 cores and 128GB of ram so I don't think it's a memory issue.

Before I dive in and poke around and/or find a workaround, has anyone else encountered this? 
There are a number of fixes for "RuntimeError: populate() isn't reentrant" but I don't want to update any modules if it's something else.

Thanks for any insight!

Error output:

**First load (Internal Server Error rendered to browser)**
```
[Sat Feb 25 16:43:18.547934 2023] [wsgi:error] [pid 7753:tid 140279247021824] [remote 172.26.0.1:46010] mod_wsgi (pid=7753): Failed to exec Python script file '/opt/openoni/onisite/wsgi.py'.
[Sat Feb 25 16:43:18.547967 2023] [wsgi:error] [pid 7753:tid 140279247021824] [remote 172.26.0.1:46010] mod_wsgi (pid=7753): Exception occurred processing WSGI script '/opt/openoni/onisite/wsgi.py'.
[Sat Feb 25 16:43:18.548524 2023] [wsgi:error] [pid 7753:tid 140279247021824] [remote 172.26.0.1:46010] Traceback (most recent call last):
[Sat Feb 25 16:43:18.548562 2023] [wsgi:error] [pid 7753:tid 140279247021824] [remote 172.26.0.1:46010]   File "/opt/openoni/onisite/wsgi.py", line 18, in <module>
[Sat Feb 25 16:43:18.548565 2023] [wsgi:error] [pid 7753:tid 140279247021824] [remote 172.26.0.1:46010]     application = get_wsgi_application()
[Sat Feb 25 16:43:18.548571 2023] [wsgi:error] [pid 7753:tid 140279247021824] [remote 172.26.0.1:46010]   File "/opt/openoni/ENV/lib/python3.8/site-packages/django/core/wsgi.py", line 12, in get_wsgi_application
[Sat Feb 25 16:43:18.548574 2023] [wsgi:error] [pid 7753:tid 140279247021824] [remote 172.26.0.1:46010]     django.setup(set_prefix=False)
[Sat Feb 25 16:43:18.548579 2023] [wsgi:error] [pid 7753:tid 140279247021824] [remote 172.26.0.1:46010]   File "/opt/openoni/ENV/lib/python3.8/site-packages/django/__init__.py", line 24, in setup
[Sat Feb 25 16:43:18.548582 2023] [wsgi:error] [pid 7753:tid 140279247021824] [remote 172.26.0.1:46010]     apps.populate(settings.INSTALLED_APPS)
[Sat Feb 25 16:43:18.548587 2023] [wsgi:error] [pid 7753:tid 140279247021824] [remote 172.26.0.1:46010]   File "/opt/openoni/ENV/lib/python3.8/site-packages/django/apps/registry.py", line 114, in populate
[Sat Feb 25 16:43:18.548589 2023] [wsgi:error] [pid 7753:tid 140279247021824] [remote 172.26.0.1:46010]     app_config.import_models()
[Sat Feb 25 16:43:18.548594 2023] [wsgi:error] [pid 7753:tid 140279247021824] [remote 172.26.0.1:46010]   File "/opt/openoni/ENV/lib/python3.8/site-packages/django/apps/config.py", line 301, in import_models
[Sat Feb 25 16:43:18.548597 2023] [wsgi:error] [pid 7753:tid 140279247021824] [remote 172.26.0.1:46010]     self.models_module = import_module(models_module_name)
[Sat Feb 25 16:43:18.548602 2023] [wsgi:error] [pid 7753:tid 140279247021824] [remote 172.26.0.1:46010]   File "/usr/lib/python3.8/importlib/__init__.py", line 127, in import_module
[Sat Feb 25 16:43:18.548605 2023] [wsgi:error] [pid 7753:tid 140279247021824] [remote 172.26.0.1:46010]     return _bootstrap._gcd_import(name[level:], package, level)
[Sat Feb 25 16:43:18.548610 2023] [wsgi:error] [pid 7753:tid 140279247021824] [remote 172.26.0.1:46010]   File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
[Sat Feb 25 16:43:18.548615 2023] [wsgi:error] [pid 7753:tid 140279247021824] [remote 172.26.0.1:46010]   File "<frozen importlib._bootstrap>", line 991, in _find_and_load
[Sat Feb 25 16:43:18.548620 2023] [wsgi:error] [pid 7753:tid 140279247021824] [remote 172.26.0.1:46010]   File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
[Sat Feb 25 16:43:18.548626 2023] [wsgi:error] [pid 7753:tid 140279247021824] [remote 172.26.0.1:46010]   File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
[Sat Feb 25 16:43:18.548631 2023] [wsgi:error] [pid 7753:tid 140279247021824] [remote 172.26.0.1:46010]   File "<frozen importlib._bootstrap_external>", line 848, in exec_module
[Sat Feb 25 16:43:18.548637 2023] [wsgi:error] [pid 7753:tid 140279247021824] [remote 172.26.0.1:46010]   File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
[Sat Feb 25 16:43:18.548642 2023] [wsgi:error] [pid 7753:tid 140279247021824] [remote 172.26.0.1:46010]   File "/opt/openoni/core/models.py", line 15, in <module>
[Sat Feb 25 16:43:18.548645 2023] [wsgi:error] [pid 7753:tid 140279247021824] [remote 172.26.0.1:46010]     from lxml import etree
[Sat Feb 25 16:43:18.548656 2023] [wsgi:error] [pid 7753:tid 140279247021824] [remote 172.26.0.1:46010] ImportError: Interpreter change detected - this module can only be loaded into one interpreter per process.
```

**Second load (page reload in browser - Internal Server Error)**

```
[Sat Feb 25 16:48:22.646282 2023] [wsgi:error] [pid 7753:tid 140279129523968] [remote 172.26.0.1:40714] mod_wsgi (pid=7753): Failed to exec Python script file '/opt/openoni/onisite/wsgi.py'.
[Sat Feb 25 16:48:22.646314 2023] [wsgi:error] [pid 7753:tid 140279129523968] [remote 172.26.0.1:40714] mod_wsgi (pid=7753): Exception occurred processing WSGI script '/opt/openoni/onisite/wsgi.py'.
[Sat Feb 25 16:48:22.646436 2023] [wsgi:error] [pid 7753:tid 140279129523968] [remote 172.26.0.1:40714] Traceback (most recent call last):
[Sat Feb 25 16:48:22.646473 2023] [wsgi:error] [pid 7753:tid 140279129523968] [remote 172.26.0.1:40714]   File "/opt/openoni/onisite/wsgi.py", line 18, in <module>
[Sat Feb 25 16:48:22.646477 2023] [wsgi:error] [pid 7753:tid 140279129523968] [remote 172.26.0.1:40714]     application = get_wsgi_application()
[Sat Feb 25 16:48:22.646485 2023] [wsgi:error] [pid 7753:tid 140279129523968] [remote 172.26.0.1:40714]   File "/opt/openoni/ENV/lib/python3.8/site-packages/django/core/wsgi.py", line 12, in get_wsgi_application
[Sat Feb 25 16:48:22.646489 2023] [wsgi:error] [pid 7753:tid 140279129523968] [remote 172.26.0.1:40714]     django.setup(set_prefix=False)
[Sat Feb 25 16:48:22.646496 2023] [wsgi:error] [pid 7753:tid 140279129523968] [remote 172.26.0.1:40714]   File "/opt/openoni/ENV/lib/python3.8/site-packages/django/__init__.py", line 24, in setup
[Sat Feb 25 16:48:22.646499 2023] [wsgi:error] [pid 7753:tid 140279129523968] [remote 172.26.0.1:40714]     apps.populate(settings.INSTALLED_APPS)
[Sat Feb 25 16:48:22.646506 2023] [wsgi:error] [pid 7753:tid 140279129523968] [remote 172.26.0.1:40714]   File "/opt/openoni/ENV/lib/python3.8/site-packages/django/apps/registry.py", line 83, in populate
[Sat Feb 25 16:48:22.646510 2023] [wsgi:error] [pid 7753:tid 140279129523968] [remote 172.26.0.1:40714]     raise RuntimeError("populate() isn't reentrant")
[Sat Feb 25 16:48:22.646525 2023] [wsgi:error] [pid 7753:tid 140279129523968] [remote 172.26.0.1:40714] RuntimeError: populate() isn't reentrant
```

**Third load (page reload in browser - Internal Server Error)**

```
[Sat Feb 25 16:54:56.683026 2023] [wsgi:error] [pid 7753:tid 140279247021824] [remote 172.26.0.1:34814] mod_wsgi (pid=7753): Failed to exec Python script file '/opt/openoni/onisite/wsgi.py'.
[Sat Feb 25 16:54:56.683058 2023] [wsgi:error] [pid 7753:tid 140279247021824] [remote 172.26.0.1:34814] mod_wsgi (pid=7753): Exception occurred processing WSGI script '/opt/openoni/onisite/wsgi.py'.
[Sat Feb 25 16:54:56.683200 2023] [wsgi:error] [pid 7753:tid 140279247021824] [remote 172.26.0.1:34814] Traceback (most recent call last):
[Sat Feb 25 16:54:56.683248 2023] [wsgi:error] [pid 7753:tid 140279247021824] [remote 172.26.0.1:34814]   File "/opt/openoni/onisite/wsgi.py", line 18, in <module>
[Sat Feb 25 16:54:56.683254 2023] [wsgi:error] [pid 7753:tid 140279247021824] [remote 172.26.0.1:34814]     application = get_wsgi_application()
[Sat Feb 25 16:54:56.683266 2023] [wsgi:error] [pid 7753:tid 140279247021824] [remote 172.26.0.1:34814]   File "/opt/openoni/ENV/lib/python3.8/site-packages/django/core/wsgi.py", line 12, in get_wsgi_application
[Sat Feb 25 16:54:56.683271 2023] [wsgi:error] [pid 7753:tid 140279247021824] [remote 172.26.0.1:34814]     django.setup(set_prefix=False)
[Sat Feb 25 16:54:56.683281 2023] [wsgi:error] [pid 7753:tid 140279247021824] [remote 172.26.0.1:34814]   File "/opt/openoni/ENV/lib/python3.8/site-packages/django/__init__.py", line 24, in setup
[Sat Feb 25 16:54:56.683286 2023] [wsgi:error] [pid 7753:tid 140279247021824] [remote 172.26.0.1:34814]     apps.populate(settings.INSTALLED_APPS)
[Sat Feb 25 16:54:56.683296 2023] [wsgi:error] [pid 7753:tid 140279247021824] [remote 172.26.0.1:34814]   File "/opt/openoni/ENV/lib/python3.8/site-packages/django/apps/registry.py", line 83, in populate
[Sat Feb 25 16:54:56.683302 2023] [wsgi:error] [pid 7753:tid 140279247021824] [remote 172.26.0.1:34814]     raise RuntimeError("populate() isn't reentrant")
[Sat Feb 25 16:54:56.683324 2023] [wsgi:error] [pid 7753:tid 140279247021824] [remote 172.26.0.1:34814] RuntimeError: populate() isn't reentrant
Exception ignored in: <function Local.__del__ at 0x7f9546e18af0>
Traceback (most recent call last):
  File "/opt/openoni/ENV/lib/python3.8/site-packages/asgiref/local.py", line 94, in __del__
NameError: name 'TypeError' is not defined
Exception ignored in: <function Local.__del__ at 0x7f9546e18af0>
Traceback (most recent call last):
  File "/opt/openoni/ENV/lib/python3.8/site-packages/asgiref/local.py", line 94, in __del__
NameError: name 'TypeError' is not defined
```

The application then _usually_ returns after the 4th page reload in the browser and no further errors occur for several hours.

**Comment by birwin79** [comment]
As a workaround, I set up a cron job to reboot the web server container every 24 hours.  So far so good, no errors for over 48 hours.  I'll keep looking into the error logs and report back with any updates.

**Comment by birwin79** [comment]
Hello @techgique.  I'm still researching the error.  I tried deleting and recreating the web container and still got the occasional crash/500 Internal Server Error.   It did crash on me during use for the first time yesterday with the following log event:

```
mod_wsgi (pid=1645): Failed to exec Python script file '/opt/openoni/onisite/wsgi.py'.
mod_wsgi (pid=1645): Exception occurred processing WSGI script '/opt/openoni/onisite/wsgi.py'.
Traceback (most recent call last):
  File "/opt/openoni/onisite/wsgi.py", line 18, in <module>
    application = get_wsgi_application()
  File "/opt/openoni/ENV/lib/python3.8/site-packages/django/core/wsgi.py", line 12, in get_wsgi_application
    django.setup(set_prefix=False)
  File "/opt/openoni/ENV/lib/python3.8/site-packages/django/__init__.py", line 24, in setup
    apps.populate(settings.INSTALLED_APPS)
  File "/opt/openoni/ENV/lib/python3.8/site-packages/django/apps/registry.py", line 114, in populate
    app_config.import_models()
  File "/opt/openoni/ENV/lib/python3.8/site-packages/django/apps/config.py", line 301, in import_models
    self.models_module = import_module(models_module_name)
  File "/usr/lib/python3.8/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 848, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/opt/openoni/core/models.py", line 15, in <module>
    from lxml import etree
ImportError: Interpreter change detected - this module can only be loaded into one interpreter per process.
```

The application returned on its own with a couple of page reloads, as anticipated.  

I think it might be related to an ongoing, silent error that keeps appearing in the log several times a day:

```
NameError: name 'TypeError' is not defined
Exception ignored in: <function Local.__del__ at 0x7f9546e19af0>
Traceback (most recent call last):
  File "/opt/openoni/ENV/lib/python3.8/site-packages/asgiref/local.py", line 94, in __del__
```

Research into this error has many avenues on Stack Overflow, which will take some time to narrow down. 

I think my mod_wsgi is current and compiled for the correct version of Python, given the output from ldd in the modules:

```
/usr/lib/apache2/modules# ldd mod_wsgi.so
        linux-vdso.so.1 (0x00007ffdc6db0000)
        libpython3.8.so.1.0 => /lib/x86_64-linux-gnu/libpython3.8.so.1.0 (0x00007fdcac9eb000)
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007fdcac9c8000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fdcac7d6000)
        libexpat.so.1 => /lib/x86_64-linux-gnu/libexpat.so.1 (0x00007fdcac7a8000)
        libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007fdcac78c000)
        libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007fdcac786000)
        libutil.so.1 => /lib/x86_64-linux-gnu/libutil.so.1 (0x00007fdcac77f000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007fdcac630000)
        /lib64/ld-linux-x86-64.so.2 (0x00007fdcacf87000)
```

Not sure if it matters, but there are two versions of the .so module files for mod_wsgi in the same folder:
/usr/lib/apache2/modules/
```
mod_wsgi.so
mod_wsgi.so-3.8
```

I set up another server with a fresh v1.1.0 install of open-oni and can confirm the intermittent issue persists there.  I would like to know if it has something to do with my Docker environment since no one else is encountering this.  I'm using Docker version 20.10.23, build 7155243, Portainer Enterprise, and Nginx Proxy Manager, and the container data & volumes are mapped to the host on a 2nd, larger nvme drive.

I'll keep researching these errors and report back with my progress!   In the meantime, I feel it is stable enough for my organization to proceed with our plans of going live with our initial batches this Friday.  _We wanted to launch the project to coincide with the anniversary of our local newspaper._  I had some beta testers try it out for a few weeks, and they didn't report any issues. 🤞

**Comment by birwin79** [comment]
I enabled the production mode of the WSGIDaemonProcess (it was previously set as Development with a maximum of 5 requests) in `/open-oni/docker/apache/openoni.conf`

It's been a few hours, and I have not seen any TypeErrors or lxml ImportErrors. 👍 (_There would typically be several errors in the log by now._)  Perhaps I set the maximum-requests and processes too low in the development config. 🤔 

I'll keep monitoring the logs and report back!

**Comment by birwin79** [comment]
No more crashes occurred after putting WSGIDaemonProcess into production mode in the apache config.
`/open-oni/docker/apache/openoni.conf`
A simple fix and hopefully others can reference this if they encounter similar issues.
Thanks!

**Comment by techgique** [comment]
@birwin79, glad to hear you're not having crashes and errors anymore with the production Apache config! And thanks for the detailed sharing so others can reference this thread if they encounter similar issues.

Also congrats on the anniversary of your local newspaper! Are your initial batches public? We like to link to projects using Open ONI :smile:

**Comment by birwin79** [comment]
@techgique  Thanks!  Our organization (The Vacaville Heritage Council) has been very impressed with the software and performance.  We launched our ONI installation in time to celebrate the 140th Anniversary of the Vacaville Reporter! 

Feel free to link to our site!  https://www.vacavilleheritagecouncil.org/oni/

We just started the OCR process in November and have launched the first 50 years.  We are scanning and bagging it all in-house with volunteers and intend to upload new batches regularly.  I created some PHP scripts to handle batch processing and a little web application to assemble the issues and tag them to the correct dates.  _We also built a function to quickly change the LCCN and other metadata in the event these batches were uploaded to NDNP in the future_

It has been an exciting project!  We thank you and the Open ONI team for making this possible!
